Unbreak docker setup for debian/ubuntu due to pip path issue

genodeftest opened the pull request:
This should help debugging exaile/exaile-testimg/#7

Building these changes is expected to fail the same way as #7.

Without these changes, all debian-based images will fail with these lines:
```
Successfully installed pip-10.0.0 setuptools-39.0.1 wheel-0.31.0
Traceback (most recent call last):
  File "/usr/bin/pip", line 9, in <module>
    from pip import main
ImportError: cannot import name main
The command '/bin/sh -c true   && export DEBIAN_FRONTEND=noninteractive   && apt-get update   && apt-get install -y --no-install-recommends     python-pytest     python-mox3     python-pip     python-setuptools     python-wheel     make   && rm -rf /var/lib/apt/lists/*   && pip install -U pip setuptools wheel   && pip install sphinx sphinx-rtd-theme' returned a non-zero code: 1
```
[Link to one failed travis build](https://travis-ci.org/exaile/exaile-testimg/builds/366711634)

This is pretty weird: Once I split up the `RUN` command to multiple lines, the setup works fine. I.e. so *somehow*, `RUN cmd1 && cmd2` is different from 

    RUN cmd1
    
    RUN cmd2

The [Docker documentation](https://docs.docker.com/engine/reference/builder/#run) only says that each argument of `RUN` is being invoked on a different shell so it looks like `pip install -U pip setuptools` does not like to be executed after any other command in a shell.

genodeftest commented:
So *somehow*, `RUN cmd1 && cmd2` is different from 

    RUN cmd1
    
    RUN cmd2

sjohannes commented:
I think what's happening is that `pip install -U pip` causes a newer version of Pip to be installed as `/usr/local/bin/pip`, in addition to the one installed by the package manager `/usr/bin/pip`. However, Bash already cached `pip` to `/usr/bin/pip` and the next time we run `pip` it uses that old one.  Something like that. I think explicitly running `/usr/local/bin/pip` would fix this, but what you have also works.

---

Aside 1: Regarding `RUN cmd1 \n RUN cmd2`: it's more than just running a different shell; it's also creating a new docker layer. Probably a tiny bit slower than doing everything in one RUN command but not a big deal.

Aside 2: For better debuggability, I usually use `RUN set -x && cmd1 && cmd2 && ...`. This causes Bash to print `+cmd1`, `+cmd2`, and so on before running them so you can see which command failed.

genodeftest commented:
> I think what's happening is that `pip install -U pip` causes a newer version of Pip to be installed as `/usr/local/bin/pip`, in addition to the one installed by the package manager `/usr/bin/pip`. However, Bash already cached `pip` to `/usr/bin/pip` and the next time we run `pip` it uses that old one. Something like that. I think explicitly running `/usr/local/bin/pip` would fix this, but what you have also works.

I don't follow completely. I understand that bash may do some caching. At the point where `pip install -U pip` is being installed, there is no `/usr/local/bin/pip` as `pip` hasn't been executed before that.

sjohannes commented:
> I don't follow completely. I understand that bash may do some caching. At the point where pip install -U pip is being installed, there is no /usr/local/bin/pip as pip hasn't been executed before that.

Yep, so `pip install -U pip` runs `/usr/bin/pip` and creates `/usr/local/bin/pip`. This runs successfully, as the log shows:

```
Successfully installed pip-10.0.0 setuptools-39.0.1 wheel-0.31.0
```

The failure then happens on the next command (`pip install sphinx sphinx-rtd-theme`). We ideally want to use the new `/usr/local/bin/pip`, but instead Bash chooses `/usr/bin/pip`:

```
Traceback (most recent call last):
  File "/usr/bin/pip", line 9, in <module>  # <-- wrong pip
    from pip import main
ImportError: cannot import name main
```

This should theoretically still work (it should just use the old Pip normally). But I think at that point Pip itself gets confused and is probably mixing up its internal libraries in `/usr/lib` and `/usr/local/lib`.

genodeftest committed:
Fix pip invocation

After upgrading pip through pip, invoking `pip` still invokes the old
`/usr/bin/pip` instead of `/usr/local/bin/pip`, which results in breaking
the following `pip install` commands. Work around this issue by invoking
`pip install` in a separate shell afterwards.

genodeftest commented:
Ok, understood. Should work fine now.

virtuald commented:
Merging shortly.
